src/components/TodoItem.tsx

```diff
@@ -0,0 +1,134 @@
+/* eslint-disable jsx-a11y/control-has-associated-label */
+import { useEffect, useRef, useState } from 'react';
+import classNames from 'classnames';
+import { Todo } from '../types/Todo';
+
+interface Props {
+  todoItem: Todo,
+  onDelete?: (id: number) => void,
+  onUpdateTodo?: (updatedTodo: Todo) => void;
+  loader: boolean,
+}
+
+export const TodoItem: React.FC<Props> = ({
+  todoItem,
+  onDelete = () => { },
+  onUpdateTodo,
+  loader,
+}) => {
+  const {
+    id, title, completed, userId,
+  } = todoItem;
+  const [hover, setHover] = useState(false);
+  const [editing, setEditing] = useState(false);
+  const [newTitle, setNewTitle] = useState(title);
+
+  const inputRef = useRef<HTMLInputElement>(null);
+
+  useEffect(() => {
+    if (editing && inputRef.current) {
+      inputRef.current.focus();
+    }
+  }, [editing]);
+
+  function toggleCompleted() {
+    if (onUpdateTodo) {
+      onUpdateTodo({
+        title,
+        id,
+        userId,
+        completed: !completed,
+      });
+    }
+  }
+
+  const onSubmit = async () => {
+    if (title === newTitle) {
+      setEditing(false);
+
+      return;
+    }
+
+    if (onUpdateTodo) {
+      await onUpdateTodo({
+        ...todoItem,
+        title: newTitle.trim(),
+      });
+    }
+
+    setEditing(false);
+  };
+
+  return (
+    <div
+      data-cy="Todo"
+      className={classNames(
+        'todo', { completed },
+      )}
+      key={id}
+      onMouseEnter={() => setHover(true)}
+      onMouseLeave={() => setHover(false)}
+      onDoubleClick={() => setEditing(true)}
+    >
+      <label className="todo__status-label">
+        <input
+          data-cy="TodoStatus"
+          type="checkbox"
+          className="todo__status"
+          checked={completed}
+          onChange={toggleCompleted}
+        />
+      </label>
+
+      {
+        editing
+          ? (
+            <form
+              onSubmit={onSubmit}
+              onBlur={onSubmit}
+            >
+              <input
+                ref={inputRef}
+                data-cy="TodoTitleField"
+                type="text"
+                className="todo__title-field"
+                placeholder="Empty todo will be deleted"
+                value={newTitle}
+                onChange={event => setNewTitle(event.target.value)}
+                onDoubleClick={() => {
+                  setNewTitle(title);
+                  setEditing(false);
+                }}
```

Comment:
Pls move the function inside `onDoubleClick` handler to a separate const and use as a callback here